File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        # we only add an open paren if open < n
        # we only cloing paren if cllsing < open
        # if open = closed paren -> we cant sdd closing paren

        stack = []
        res = []

        def backtrack(openN, closedN):
            if closedN == openN == n:
                res.append("".join(stack))
                return


            if openN < n:
                stack.append("(")
                backtrack(openN + 1, closedN)
                stack.pop()

            if closedN < openN:
                stack.append(")")
                backtrack(openN, closedN + 1)
                stack.pop()

        backtrack(0, 0)
        return res

File: 0022-generate-parentheses/0022-generate-parentheses.java
import java.util.*;

class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        backtrack(n, n, new StringBuilder(), res);
        return res;
    }

    private void backtrack(int openLeft, int closeLeft, StringBuilder sb, List<String> res) {
        if (openLeft == 0 && closeLeft == 0) {
            res.add(sb.toString());
            return;
        }
        if (openLeft > 0) {                // you can always place '(' if any are left
            sb.append('(');
            backtrack(openLeft - 1, closeLeft, sb, res);
            sb.setLength(sb.length() - 1); // pop
        }
        if (closeLeft > openLeft) {        // only place ')' if it won't break validity
            sb.append(')');
            backtrack(openLeft, closeLeft - 1, sb, res);
            sb.setLength(sb.length() - 1); // pop
        }
    }
}
